src/main/java/org/usfirst/frc4048/common/swervev2/SwerveDrivetrain.java

```diff
@@ -0,0 +1,170 @@
+        
+        EncodedSwerveSparkMax encodedSwerveSparkMaxFL = new EncodedSwerveSparkMax(frontLeftDrive, frontLeftTurn, frontLeftCanCoder, driveVelConvFactor, drivePosConvFactor, steerPosConvFactor);
+        EncodedSwerveSparkMax encodedSwerveSparkMaxFR = new EncodedSwerveSparkMax(frontRightDrive, frontRightTurn, frontRightCanCoder, driveVelConvFactor, drivePosConvFactor, steerPosConvFactor);
+        EncodedSwerveSparkMax encodedSwerveSparkMaxBL = new EncodedSwerveSparkMax(backLeftDrive, backLeftTurn, backLeftCanCoder, driveVelConvFactor, drivePosConvFactor, steerPosConvFactor);
+        EncodedSwerveSparkMax encodedSwerveSparkMaxBR = new EncodedSwerveSparkMax(backRightDrive, backRightTurn, backRightCanCoder, driveVelConvFactor, drivePosConvFactor, steerPosConvFactor);
+        TrapezoidProfile.Constraints constraints = new TrapezoidProfile.Constraints(Constants.MAX_ANGULAR_SPEED * 4, MODULE_MAX_ANGULAR_ACCELERATION * 10);
+        PID drivePid = PID.of(Constants.DRIVE_PID_P,Constants.DRIVE_PID_I,Constants.DRIVE_PID_D);
+        PID steerPid = PID.of(Constants.STEER_PID_P,Constants.STEER_PID_I,Constants.STEER_PID_D);
+        Gain driveGain = Gain.of(Constants.DRIVE_PID_FF_V,Constants.DRIVE_PID_FF_S);
+        Gain steerGain = Gain.of(Constants.STEER_PID_FF_V,Constants.STEER_PID_FF_S);
+        frontLeft = new SparkMaxSwerveModule(encodedSwerveSparkMaxFL, drivePid,steerPid, driveGain,steerGain,constraints);
+        frontRight = new SparkMaxSwerveModule(encodedSwerveSparkMaxFR, drivePid,steerPid, driveGain,steerGain,constraints);
+        backLeft = new SparkMaxSwerveModule(encodedSwerveSparkMaxBL, drivePid,steerPid, driveGain,steerGain,constraints);
+        backRight = new SparkMaxSwerveModule(encodedSwerveSparkMaxBR, drivePid,steerPid, driveGain,steerGain,constraints);
+        frontLeftDrive.setInverted(true);
+        frontRightDrive.setInverted(false);
+        backRightDrive.setInverted(false);
+        backLeftDrive.setInverted(true);
+        navxGyro.setAngleAdjustment(gyroOffset);
+    }
+
+    @Override
+    public void periodic() {
+        navxGyroValue = getGyro();
+    }
+    public float getRoll() {
+        return navxGyro.getRoll();
+    }
+    
+    private static final double MODULE_MAX_ANGULAR_ACCELERATION =
+            2 * Math.PI; // radians per second squared
+    public void drive(double xSpeed, double ySpeed, double rot, boolean fieldRelative) {
+        SwerveModuleState[] swerveModuleStates = kinematics.toSwerveModuleStates(
+                fieldRelative
+                        ? ChassisSpeeds.fromFieldRelativeSpeeds(xSpeed, ySpeed, rot, new Rotation2d(Math.toRadians(navxGyroValue)))
+                        : new ChassisSpeeds(xSpeed, ySpeed, rot));
+        SwerveDriveKinematics.desaturateWheelSpeeds(swerveModuleStates, Constants.MAX_VELOCITY);
+        setModuleStates(swerveModuleStates);
+    }
+    public void driveRaw(double speed){
```

Comment:
What is this doing? Where is the speed directed? Maybe add documentation?

Reply:
forgot to delete this part from testing (dont)